lib/AST/ASTScopeCreation.cpp

```diff
@@ -0,0 +1,1061 @@
+  void visitDeferStmt(DeferStmt *ds, ASTScopeImpl *p, DeferredNodes &deferred) {
+    visitFuncDecl(ds->getTempDecl(), p, deferred);
+  }
+
+  void visitBraceStmt(BraceStmt *bs, ASTScopeImpl *p, DeferredNodes &deferred) {
+    if (p->areDeferredNodesInANewScope())
+      p->createSubtreeWithDeferrals<BraceStmtScope>(deferred, bs);
+    else
+      p->createSubtree<BraceStmtScope>(bs);
+  }
+
+  void visitPatternBindingDecl(PatternBindingDecl *patternBinding,
+                               ASTScopeImpl *p, DeferredNodes &deferred) {
+    p->createAttachedPropertyDelegateScope(patternBinding);
+    // deferred will contain any nodes that need to be put into subscopes.
+    // In a type decl body, there won't be any.
+    // The comment for \c AbstractPatternEntryScope.
+    const bool isInTypeDecl =
+        p->getDecl() && (isa<NominalTypeDecl>(p->getDecl().get()) ||
+                              isa<ExtensionDecl>(p->getDecl().get()));
+    if (isInTypeDecl)
+      p->createScopesForPatternBindingInATypeDecl(patternBinding);
+    else
+      p->createScopesForPatternBindingInAFunction(patternBinding,
+                                                       deferred);
+  }
+
+  void visitReturnStmt(ReturnStmt *rs, ASTScopeImpl *p,
+                       DeferredNodes &deferred) {
+    if (rs->hasResult())
+      visitExpr(rs->getResult(), p, deferred);
+  }
+
+  void visitWhileStmt(WhileStmt *ws, ASTScopeImpl *p, DeferredNodes &deferred) {
+    if (ws->getCond().empty()) {
+      visit(ws->getBody(), p, deferred);
+      return;
+    }
+    p->createSubtree<WhileConditionalClauseScope>(ws, 0);
+  }
+
+  void visitExpr(Expr *expr, ASTScopeImpl *p, DeferredNodes &deferred) {
+    if (!expr)
+      return;
+    p->widenSourceRangeForIgnoredASTNode(expr);
+    p->addChildrenForCapturesAndClosuresIn(expr);
+  }
+};
+} // namespace ast_scope
+} // namespace swift
+
+#pragma mark dispatch-and-create
+
+void ASTScopeImpl::dispatchAndCreateAll(DeferredNodes &deferred) {
+  deferred.setLastAdopter(this); // in case we come back here later after adding
+                                 // more Decls to the SourceFile
+  while (!deferred.empty()) {
+    Optional<ASTNode> node = deferred.popNextDeferredNodeForAdoptionBy(this);
+    if (!node)
+      return;
+    if (auto declToScope = node->dyn_cast<Decl *>())
+      dispatchAndCreateIfNeeded(declToScope, deferred);
+    else if (auto stmtToScope = node->dyn_cast<Stmt *>())
+      dispatchAndCreateIfNeeded(stmtToScope, deferred);
+    else if (auto exprToScope = node->dyn_cast<Expr *>())
+      dispatchAndCreateIfNeeded(exprToScope, deferred);
+    else
+      llvm_unreachable("Impossible case");
+  }
+}
+
+template <typename StmtExpr>
+void ASTScopeImpl::dispatchAndCreateIfNeeded(StmtExpr *se,
+                                             DeferredNodes &deferred) {
+  if (se)
+    ScopeCreator().visit(se, this, deferred);
+}
+void ASTScopeImpl::dispatchAndCreateIfNeeded(Decl *d, DeferredNodes &deferred) {
```

Comment:
Can you make sure to put spaces in between things like this?

Reply:
Yes, a space there is good, thanks. Other places, I'm not so sure:
```
ASTScopeImpl *
ExtensionScope::createTrailingWhereClauseScope(ASTScopeImpl *parent) {
  return parent->createSubtree2D<ExtensionScope, GTXWherePortion>(decl);
}
ASTScopeImpl *
NominalTypeScope::createTrailingWhereClauseScope(ASTScopeImpl *parent) {
  return parent->createSubtree2D<NominalTypeScope, GTXWherePortion>(decl);
}
ASTScopeImpl *
TypeAliasScope::createTrailingWhereClauseScope(ASTScopeImpl *parent) {
  return parent->createSubtree2D<TypeAliasScope, GTXWherePortion>(decl);
}
```
Here, I could use a `define`, but it seems not quite an improvement for such a small number. So, I go with tight repetition.